Johnfreeman/daq deliverables issue105 standard image

jcfreeman2 opened the pull request.

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: add a draft of the image one layer
between the base image and the image with the Python packages

Note that this isn't a literal union of all the packages currently
getting yum-installed in the Dockerfiles in this repo. This is because
some packages are already available in the base image (e.g., bzip2 or
openssl), or in the case of the oracle client software, it doesn't
appear to be available in the repos reachable from this image. It _is_
available in the cern/cc7-base image, but the problem there is that
there's no Python version past 3.6 which can be installed by yum
(though to be fair, it could be installed manually if push came to
shove)

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: update base image in cern/alma9-base and
add requirements.txt

For the time being will just use the Python which comes with cern/alma9-base
(v3.9.16), as this appears to be tied in to yum itself via the dnf Python package.
The Python packages in the requirements.txt are the union of the packages used
for the current microservices on the develop branch, minus cx-Oracle both because
Oracle use is uncertain moving forward and because it requires a dataclasses
version (0.8) which appears to be unavailable according to pip.

The versions of the Python packages are preferentially (1) the version found
in our nightly release, (2) the newest version across the microservice Python
packages, (3) in the case where no version was originally provided, the default
version which pip3 installed.

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: add version of the standard Docker image which includes Oracle

A couple of quick points:

-This needed to be based on cern/cc7-base since as far as I'm aware this is the only CERN Docker image which provides a repo
 (cernonly) containing the Oracle client software

-Python needed to be installed manually since Python 3.6 is the newest Python version available in the yum repos

-The Dockerfile is a bit hacky in places (softlinking, e.g.) but can be brushed up

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: standardize use of entrypoint.sh scripts and add cx-Oracle Python package as standard, whether or not you actually have Oracle client software installed

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: update the Dockerfile for the Oracle-inclusive image so that it builds Python based on OpenSSL v1.1.1; otherwise importing the ssl Python module won't work and (among other things) importing the requests Python module will fail

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: Dockerfile_w_oracle wasn't committed correctly in previous commit; fixed

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: remove ~2.5 unnecessary GB from the image

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: install the Oracle client software from the website Alessandro pointed me to in the cern/alma9-based Dockerfile; rename Dockerfile_w_oracle to Dockerfile_cc7 to reflect what the file now stands out for (and consider for deletion in the future)

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: implement some changes to the Dockerfile suggested by Pat and Bonnie

Specifically:
-Reduce the number of layers in the standard microservices image
-Move elisa_client_api installation out of the logbook's entrypoint script and into the Dockerfile
-Post a note suggesting that in the future we not run as root inside the microservices container

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: remove now-obsolete individual microservice Dockerfile and requirements.txt

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: update and standardize the documentation

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: add support for Google's Protobuf, and generate Python code from the *.proto file in ers for Python modules (e.g. ERSSubscriber in erskafka) to use

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: just in case we need to fall back on the Centos7-based image, add support for Protobuf there as well

jcfreeman2 commented:
This PR is the followup to https://github.com/DUNE-DAQ/microservices/pull/61, which I closed after a discussion about Marco's microservice needs. Since that PR, I've updated the standard image to include Protobuf and the Python code it generates off of ers's `schema/ers/issue.proto` so that you can `import ers.issue_pb2` in your Python scripts. 

Instructions on how to test this PR are in https://github.com/DUNE-DAQ/microservices/tree/johnfreeman/daq-deliverables_issue105_standard_image#readme . Please note that while they refer to the `Alma9`-based microservices image, we have a fallback `Centos7`-based image, `ghcr.io/dune-daq/microservices_cc7:ec76`, which we can also test.

mroda88 commented:
Unfortunately `curl` and `git clone` don't work from the entrypoint so this is not enough for the new ERS microservice.

plasorak commented:
For me, the base image should have the required functionality to get :
 - The base image should be relatively "generic". I.e. it should just provide specific packages that we commonly rely on (oracle, protobuf, git, etc.).
 - Microservices should start from that image (`FROM generic_daq_microservice:A9.v1.0.0`, for example at the top of their `Dockerfile`) and install the code that they need. So for example for the ERS streamer:
```
FROM generic_daq_microservice:A9.v1.0.0
ARGS ERSVERSION=dunedaq-v4.1.1

RUN git clone https://github.com/DUNE-DAQ/ers.git -b $ERSVERSION

# The following script could be created in the base image:
RUN create_pypkg_protobuf ers/schema/ers ersschema 
RUN cd ers && pip install . # or whatever is needed to install the ers streamer

ENTRYPOINT ["fancy-ers-streamer-db-writer"]
```

That way we ensure that no `git clone` gets used every time we start the image, and we still have a relatively generic image with common functionalities stored in the same place. This means we don't rely on network calls that can potentially fail, and on the sanity of the repo if someone accidentally modifies it destructively.

`druncschema` has an example of generating protobuf files and that can be installed with `pip install .` without interfering with the dunedaq-pyvenv: https://github.com/DUNE-DAQ/druncschema/tree/develop

alessandrothea commented on standard/Dockerfile:
```diff
@@ -0,0 +1,39 @@
+# protoc-24.3-linux-x86_64.zip is the latest zipfile available as of Sep-15-2023
+# See also https://grpc.io/docs/protoc-installation/#install-pre-compiled-binaries-any-os
+
+RUN curl -LO https://github.com/protocolbuffers/protobuf/releases/download/v24.3/protoc-24.3-linux-x86_64.zip && \
```
isn't protobuf pip-installable?

alessandrothea reviewed (commented):
See questions in the code

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: refactor organization of microservices

This involves a few changes, including:
-A general entrypoint.sh script, which will look at the MICROSERVICES environment variable to call a specific microservices's entrypoint.sh

-Two images, one (meant to be called ghcr.io/dune-daq/microservices_dependencies:latest) which
 contains the packages and files the microservices depend on

-...and another image which inherits from this image but additionally contains the microservice repo
 itself, as discussed with Alessandro on the software coordination Slack channel

-Routine checking in the microservice-specific entrypoint.sh scripts that the needed environment
 variables are set

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: update the documentation to reflect the refactor

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: Update README.md

jcpunk committed:
Initial attempt at kubernetes deployments for unified image

jcpunk committed:
Simplify altering kafka config

jcpunk committed:
Allow customization of kafka cluster

Conflicts:
	ers-dbwriter/dbwriter.py

jcpunk committed:
Update for new var

jcpunk committed:
Fixup name

jcfreeman2 committed:
JCF: Update README.md

jcpunk committed:
Fix var interpolation

jcfreeman2 committed:
JCF: Update README.md

jcfreeman2 committed:
JCF: DUNE-DAQ/daq-deliverables#105: as a finishing touch have the final layer refer to the develop branch, rather than this feature branch (since we're about to merge)

jcfreeman2 commented:
Just had a meeting with Ron, Bonnie and Pat; will merge into develop
